### PR TITLE
Kotlin daemon JVM args - clearing up inheritance and defaults

### PR DESCRIPTION
**Pull request by kaeawc**
I filed [KT-71564](https://youtrack.jetbrains.com/issue/KT-71564/Kotlin-Daemon-JVM-args-behavior-does-not-match-documentation) the other day because my previous understanding (and that of some of the AndroidX team) was that all Gradle JVM args would be inherited if no Kotlin JVM args were specified. I did some digging in the Kotlin Gradle Plugin and Kotlin Daemon sources to find how that process works and think this should clear everything up.

### CODE REVIEW

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 1]
```diff
@@ -265,7 +279,23 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+
+* -Xmx
+* -XX:MaxMetaspaceSize
+* -XX:ReservedCodeCacheSize (Defaults to 320m if not inherited and unspecified)
+
+The Kotlin daemon also has a number of defaults:
+* -XX:UseParallelGC (if no other garbage collector was set)
+* -ea
+* -XX:+UseCodeCacheFlushing
+* -XX:+HeapDumpOnOutOfMemoryError
```
Comment (ALikhachev) [inline docs/topics/gradle/gradle-compilation-and-caches.md:293]: This one have to be configured by your JVM or some other implicit configuration like the `JAVA_TOOL_OPTIONS` environment variable. That's not what does KGP configure by default.

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 2]
```diff
@@ -265,7 +279,23 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+* -Djava.awt.headless=true
+* -D{java.servername.property}={localhostip}
+* --add-exports=java.base/sun.nio.ch=ALL-UNNAMED (Only for JDK 16+)
+
+You can use a tool like [VisualVM](https://visualvm.github.io/) to verify what the actual settings of a running JVM process like the Kotlin daemon are.
```
Comment (ALikhachev) [inline docs/topics/gradle/gradle-compilation-and-caches.md:298]: Could you rephrase this as 
```suggestion
The list of defaults is considered an implementation detail of the Kotlin daemon and may vary between versions. You can use a tool like [VisualVM](https://visualvm.github.io/) to verify what the actual settings of a running JVM process like the Kotlin daemon are.
```

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 3]
```diff
@@ -163,12 +163,22 @@ Each of the following ways to set arguments overrides the ones that came before
...
 
-If nothing is specified, the Kotlin daemon inherits arguments from the Gradle daemon. For example, in the `gradle.properties` file:
+The Kotlin daemon inherits a specific set of arguments from the Gradle daemon and then overwrites them with JVM args specified for the Kotlin daemon. For example, in the `gradle.properties` file:
 
 ```none
-org.gradle.jvmargs=-Xmx1500m -Xms500m
+org.gradle.jvmargs=-Xmx1500m -Xms500m -XX:MaxMetaspaceSize=1g
 ```
 
+would then result in the Kotlin daemon's JVM args being:
```
Comment (ALikhachev) [inline docs/topics/gradle/gradle-compilation-and-caches.md:172]:
```suggestion
would then result in the Kotlin daemon's JVM args being similar to:
```

**Review by ALikhachev — CHANGES_REQUESTED** [review]

File: docs/topics/gradle/gradle-compilation-and-caches.md (hunk 1 above)
Comment (kaeawc) [reply docs/topics/gradle/gradle-compilation-and-caches.md:293]: Ah, my mistake, will remove

**Comment by kaeawc** [comment]
@sarahhaggarty is there anything else you'd want me to change in this PR?

**Review by ALikhachev — APPROVED** [review]

**Comment by kaeawc** [comment]
@sarahhaggarty @danil-pavlov just wanted to bump this issue, what might next steps be?

**Comment by sarahhaggarty** [comment]
Hi @kaeawc!
Sorry for the slow response to your PR. I'm catching up after our recent migration to writerside. I'll come back with any comments within the next two weeks.

**Comment by kaeawc** [comment]
Thanks and all good. 😄

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 4]
```diff
@@ -163,12 +163,22 @@ Each of the following ways to set arguments overrides the ones that came before
 
 #### Gradle daemon arguments inheritance
 
-If nothing is specified, the Kotlin daemon inherits arguments from the Gradle daemon. For example, in the `gradle.properties` file:
+The Kotlin daemon inherits a specific set of arguments from the Gradle daemon and then overwrites them with JVM args specified for the Kotlin daemon. For example, in the `gradle.properties` file:
```
Comment (sarahhaggarty) [inline docs/topics/gradle/gradle-compilation-and-caches.md:166]:
```suggestion
By default, the Kotlin daemon inherits a specific set of arguments from the Gradle daemon but overwrites them with any 
JVM arguments specified directly for the Kotlin daemon. For example, if you add the following JVM arguments in the `gradle.properties` file:
```

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 5]
```diff
@@ -163,12 +163,22 @@ Each of the following ways to set arguments overrides the ones that came before
...
 
 ```none
-org.gradle.jvmargs=-Xmx1500m -Xms500m
+org.gradle.jvmargs=-Xmx1500m -Xms500m -XX:MaxMetaspaceSize=1g
 ```
 
+would then result in the Kotlin daemon's JVM args being similar to:
```
Comment (sarahhaggarty) [inline docs/topics/gradle/gradle-compilation-and-caches.md:172]:
```suggestion
These arguments are then added to the Kotlin daemon's JVM arguments:
```

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 6]
```diff
@@ -163,12 +163,22 @@ Each of the following ways to set arguments overrides the ones that came before
...
+
+```none
+-Xmx1500m -XX:ReservedCodeCacheSize=320m -XX:MaxMetaspaceSize=1g -XX:UseParallelGC -ea -XX:+UseCodeCacheFlushing -XX:+HeapDumpOnOutOfMemoryError -Djava.awt.headless=true -Djava.rmi.server.hostname=127.0.0.1 --add-exports=java.base/sun.nio.ch=ALL-UNNAMED
+```
+
+> Learn more about [Kotlin daemon's behavior with JVM arguments](#kotlin-daemon-s-behavior-with-jvm-arguments).
```
Comment (sarahhaggarty) [inline docs/topics/gradle/gradle-compilation-and-caches.md:178]:
```suggestion
> To learn more about the Kotlin daemon's default behavior with JVM arguments, see [Kotlin daemon's behavior with JVM arguments](#kotlin-daemon-s-behavior-with-jvm-arguments).
```

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 7]
```diff
@@ -193,12 +203,16 @@ When passing arguments, follow these rules:
 
 #### kotlin.daemon.jvmargs property
 
-You can add the `kotlin.daemon.jvmargs` property in the `gradle.properties` file:
+You can add the `kotlin.daemon.jvmargs` property in the `gradle.properties` file. Note that if you do not specify `ReservedCodeCacheSize` here or in Gradle's JVM args, a default will be applied by the Kotlin Gradle plugin:
```
Comment (sarahhaggarty) [inline docs/topics/gradle/gradle-compilation-and-caches.md:206]:
```suggestion
You can add the `kotlin.daemon.jvmargs` property in the `gradle.properties` file:
```

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 8]
```diff
@@ -193,12 +203,16 @@ When passing arguments, follow these rules:
...
 
 ```none
 kotlin.daemon.jvmargs=-Xmx1500m -Xms500m
 ```
 
+```none
```
Comment (sarahhaggarty) [inline docs/topics/gradle/gradle-compilation-and-caches.md:212]:
```suggestion
Note that if you don't specify the `ReservedCodeCacheSize` argument here or in Gradle's JVM arguments, the Kotlin Gradle plugin applies a default value of `320m`:

```none
```
I think that it's better to have this sentence and it's sample separate from the first. Otherwise the second sample looks lost.

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 9]
```diff
@@ -265,7 +279,22 @@ When configuring the Kotlin daemon's JVM arguments, note that:
   > even if other requested JVM arguments are different, this daemon will be reused instead of starting a new one.
   >
   {style="note"}
-* If the `Xmx` argument is not specified, the Kotlin daemon will inherit it from the Gradle daemon.
+
+If the following arguments are not specified, the Kotlin daemon will inherit them from the Gradle daemon:
```
Comment (sarahhaggarty) [inline docs/topics/gradle/gradle-compilation-and-caches.md:283]:
```suggestion
If the following arguments aren't specified, the Kotlin daemon inherits them from the Gradle daemon:
```

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 10]
```diff
@@ -265,7 +279,22 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+
+* -Xmx
+* -XX:MaxMetaspaceSize
+* -XX:ReservedCodeCacheSize (Defaults to 320m if not inherited and unspecified)
```
Comment (sarahhaggarty) [inline docs/topics/gradle/gradle-compilation-and-caches.md:287]:
```suggestion
* `-Xmx`
* `-XX:MaxMetaspaceSize`
* `-XX:ReservedCodeCacheSize`. If not specified or inherited, the default value is `320m`.
```

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 11]
```diff
@@ -265,7 +279,22 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+
+The Kotlin daemon also has a number of defaults:
```
Comment (sarahhaggarty) [inline docs/topics/gradle/gradle-compilation-and-caches.md:289]:
```suggestion
The Kotlin daemon has the following default JVM arguments:
```

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 12]
```diff
@@ -265,7 +279,22 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+* -XX:UseParallelGC (if no other garbage collector was set)
+* -ea
+* -XX:+UseCodeCacheFlushing
+* -Djava.awt.headless=true
+* -D{java.servername.property}={localhostip}
+* --add-exports=java.base/sun.nio.ch=ALL-UNNAMED (Only for JDK 16+)
```
Comment (sarahhaggarty) [inline docs/topics/gradle/gradle-compilation-and-caches.md:295]:
```suggestion
* `-XX:UseParallelGC`. This argument is only applied if no other garbage collector is specified.
* `-ea`
* `-XX:+UseCodeCacheFlushing`
* `-Djava.awt.headless=true`
* `-D{java.servername.property}={localhostip}`
* `--add-exports=java.base/sun.nio.ch=ALL-UNNAMED`. This argument is only applied for JDK versions 16 or higher.
```

File: docs/topics/gradle/gradle-compilation-and-caches.md [hunk 13]
```diff
@@ -265,7 +279,22 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+
+The list of defaults is considered an implementation detail of the Kotlin daemon and may vary between versions. You can use a tool like [VisualVM](https://visualvm.github.io/) to verify what the actual settings of a running JVM process like the Kotlin daemon are.
```
Comment (sarahhaggarty) [inline docs/topics/gradle/gradle-compilation-and-caches.md:297]:
```suggestion
> The list of default JVM arguments for the Kotlin daemon may vary between versions. You can use a tool like [VisualVM](https://visualvm.github.io/) to check the actual settings of a running JVM process, like the Kotlin daemon.
>
{style="note"}
```

**Review by sarahhaggarty — CHANGES_REQUESTED** [review]
Thanks for your suggestions @kaeawc!
I've made some suggestions to update the text based on our style guidelines. Can you update your PR? Please let me know if you need any assistance or if something isn't clear to you.

**Review by sarahhaggarty — APPROVED** [review]

**Comment by sarahhaggarty** [comment]
Thanks for your contribution @kaeawc !
